apps/os/backend/durable-objects/estate-builds.ts

```diff
@@ -0,0 +1,392 @@
+      }
+
+      // Debug logging for ingest batches and heartbeats
+      if (logs.length === 0) {
+        logger.info("estate-builds: heartbeat received", { buildId });
+      } else {
+        logger.info("estate-builds: ingest batch", {
+          buildId,
+          batchSize: logs.length,
+          accepted,
+          lastSeqBefore: lastSeq,
+          lastSeqAfter: newLast,
+        });
+      }
+
+      return new Response(JSON.stringify({ ok: true, lastSeq: newLast }), {
+        headers: { "Content-Type": "application/json" },
+      });
+    } catch (error) {
+      logger.error("estate-builds: handleIngest error:", error);
+      return new Response("Internal error", { status: 500 });
+    }
+  }
+
+  private async handleWebSocketUpgrade(request: Request): Promise<Response> {
+    const url = new URL(request.url);
+    const estateId = url.searchParams.get("estateId") || "test";
+    const buildId = url.searchParams.get("buildId") || "test";
+
+    if (!estateId || !buildId) {
+      return new Response("Missing estateId or buildId", { status: 400 });
+    }
+
+    // Authorization is handled at the worker route via signed URLs / session checks
+
+    const pair = new WebSocketPair();
+    const [client, server] = Object.values(pair);
+    this.ctx.acceptWebSocket(server);
+    // Persist metadata across hibernations
+    try {
+      server.serializeAttachment({ estateId, buildId });
+    } catch (_err) {
+      void 0;
+    }
+
+    const watchers = this.watchersByBuildId.get(buildId) || new Set<WebSocket>();
+    watchers.add(server);
+    this.watchersByBuildId.set(buildId, watchers);
+
+    // On watcher connection, replay buffered logs from storage
+    logger.info("estate-builds: watcher connected", { buildId, watchers: watchers.size });
+    await this.replayLogsToWatcher(server, buildId);
+
+    return new Response(null, { status: 101, webSocket: client });
+  }
+
+  // No message handling needed for watcher sockets
+
+  async webSocketClose(ws: WebSocket) {
+    const attachment = (ws as any).deserializeAttachment
+      ? (ws as any).deserializeAttachment()
+      : null;
+    const buildId: string | undefined = attachment?.buildId ?? (ws as any).__buildId;
+    if (buildId) {
+      const watchers = this.watchersByBuildId.get(buildId);
+      if (watchers) {
+        watchers.delete(ws);
+        if (watchers.size === 0) this.watchersByBuildId.delete(buildId);
+      }
+    }
+  }
+
+  // Periodic alarm to detect stale builds
+  async alarm() {
+    const now = Date.now();
+    const heartbeats = await this.ctx.storage.get<Record<string, number>>("heartbeats");
+    if (!heartbeats) {
+      await this.scheduleNextAlarm();
+      return;
+    }
+
+    let changed = false;
+    for (const [buildId, last] of Object.entries(heartbeats)) {
+      if (now - last > 2 * 60_000) {
+        // 2 minutes without heartbeat → failed
+        try {
+          await this.appendLog(buildId, now, "stdout", "[BUILD FAILED]");
```

Comment:
can we have a different way of classifying timed-out/lost than failed? will help us debug